tsdb/engine/tsm1/writer.go

```diff
@@ -0,0 +1,632 @@
+	Contains(key string) bool
+
+	// ContainsValue returns true if key and time might exists in this file.  This function could
+	// return true even though the actual point does not exists.  For example, the key may
+	// exists in this file, but not have point exactly at time t.
+	ContainsValue(key string, timestamp time.Time) bool
+
+	// Entries returns all index entries for a key.
+	Entries(key string) []*IndexEntry
+
+	// Entry returns the index entry for the specified key and timestamp.  If no entry
+	// matches the key and timestamp, nil is returned.
+	Entry(key string, timestamp time.Time) *IndexEntry
+
+	// Keys returns the unique set of keys in the index.
+	Keys() []string
+
+	// Key returns the key in the index at the given postion.
+	Key(index int) (string, []*IndexEntry)
+
+	// KeyCount returns the count of unique keys in the index.
+	KeyCount() int
+
+	// Size returns the size of a the current index in bytes
+	Size() uint32
+
+	// TimeRange returns the min and max time across all keys in the file.
+	TimeRange() (time.Time, time.Time)
+
+	// KeyRange returns the min and max keys in the file.
+	KeyRange() (string, string)
+
+	// Type returns the block type of the values stored for the key.  Returns one of
+	// BlockFloat64, BlockInt64, BlockBool, BlockString.  If key does not exist,
+	// an error is returned.
+	Type(key string) (byte, error)
+
+	// MarshalBinary returns a byte slice encoded version of the index.
+	MarshalBinary() ([]byte, error)
+
+	// UnmarshalBinary populates an index from an encoded byte slice
+	// representation of an index.
+	UnmarshalBinary(b []byte) error
+}
+
+// IndexEntry is the index information for a given block in a TSM file.
+type IndexEntry struct {
+
+	// The min and max time of all points stored in the block.
+	MinTime, MaxTime time.Time
+
+	// The absolute position in the file where this block is located.
+	Offset int64
+
+	// The size in bytes of the block in the file.
+	Size uint32
+}
+
+func (e *IndexEntry) UnmarshalBinary(b []byte) error {
+	if len(b) != indexEntrySize {
+		return fmt.Errorf("unmarshalBinary: short buf: %v != %v", indexEntrySize, len(b))
+	}
+	e.MinTime = time.Unix(0, int64(btou64(b[:8])))
+	e.MaxTime = time.Unix(0, int64(btou64(b[8:16])))
+	e.Offset = int64(btou64(b[16:24]))
+	e.Size = btou32(b[24:28])
+	return nil
+}
+
+// Returns true if this IndexEntry may contain values for the given time.  The min and max
+// times are inclusive.
+func (e *IndexEntry) Contains(t time.Time) bool {
+	return (e.MinTime.Equal(t) || e.MinTime.Before(t)) &&
+		(e.MaxTime.Equal(t) || e.MaxTime.After(t))
+}
+
+func (e *IndexEntry) OverlapsTimeRange(min, max time.Time) bool {
+	return (e.MinTime.Equal(max) || e.MinTime.Before(max)) &&
+		(e.MaxTime.Equal(min) || e.MaxTime.After(min))
+}
+
+func (e *IndexEntry) String() string {
+	return fmt.Sprintf("min=%s max=%s ofs=%d siz=%d", e.MinTime.UTC(), e.MaxTime.UTC(), e.Offset, e.Size)
+}
+
+func NewDirectIndex() TSMIndex {
```

Comment:
OK, for some reason I thought indirect would be an entire replacement for direct, but you're saying only for the read case.